linkerd/protocol/http/src/main/scala/com/twitter/finagle/buoyant/linkerd/Headers.scala

```diff
@@ -1,121 +1,427 @@
+     */
+    val serverModule: Stackable[ServiceFactory[Request, Response]] =
+      new Stack.Module0[ServiceFactory[Request, Response]] {
+        val role = Stack.Role("ServerContextFilter")
+        val description = "Extracts linkerd context from http headers"
+
+        val deadline = new Deadline.ServerFilter
+        val dtab = new Dtab.ServerFilter
+
+        def make(next: ServiceFactory[Request, Response]) =
+          deadline.andThen(dtab).andThen(next)
+      }
+
+    /**
+     * A clientside stack module that injects local contextual
+     * information onto downstream requests.  Currently this includes:
+     *   - Deadline
      *
-     * Based on com.twitter.finagle.tracing.Trace.idCtx.tryUnmarshal
+     * Note that Dtabs are *not* encoded by this filter, since the
+     * HttpClientDispatcher is currently responsible for encoding the
+     * `dtab-local` header. In a future release, Dtabs will be encoded
+     * into the `l5d-ctx-dtab` header.
      *
-     * The wire format is (big-endian):
-     *   ''reqId:8 parentId:8 traceId:8 flags:8''
+     * Note that trace configuration is handled by
+     * [[HttpTraceInitializer.clientModule]].
      */
-    def get(b64: String): Try[TraceId] =
-      Try(Base64.getDecoder.decode(b64)).flatMap(TraceId.deserialize)
-
-    def get(headers: HeaderMap): Option[TraceId] =
-      for {
-        header <- headers.get(Key)
-        traceId <- get(header).toOption
-      } yield traceId
-
-    def set(headers: HeaderMap, id: TraceId): Unit = {
-      val bytes = TraceId.serialize(id)
-      val b64 = Base64.getEncoder().encodeToString(bytes)
-      val _ = headers.set(Key, b64)
+    val clientModule: Stackable[ServiceFactory[Request, Response]] =
+      new Stack.Module0[ServiceFactory[Request, Response]] {
+        val role = Stack.Role("ClientContextFilter")
+        val description = "Injects linkerd context into http headers"
+
+        // TODO use Dtab.ClientFilter once this module can replace
+        //      finagle's dtab encoding logic.
+        val deadline = new Deadline.ClientFilter
+
+        def make(next: ServiceFactory[Request, Response]) =
+          deadline.andThen(next)
+      }
+
+    val Prefix = Headers.Prefix + "ctx-"
+
+    /**
+     * The `l5d-ctx-deadline` header propagates a request
+     * deadline. Each router server may use this deadline to cancel or
+     * reject work.
+     *
+     * Each router client sets a deadline that it is at least as
+     * strict as the deadline it received. If an incoming request has
+     * a deadline, the outgoing request MUST have a
+     * deadline. Otherwise, outgoing requests MAY have a deadline.
+     */
+    object Deadline {
+      val Key = Prefix + "deadline"
+
+      def read(v: String): FDeadline = {
+        val values = v.split(' ')
+        val timestamp = Time.fromNanoseconds(values(0).toLong)
+        val deadline = Time.fromNanoseconds(values(1).toLong)
+        FDeadline(timestamp, deadline)
+      }
```

Comment:
tioli: should this return an Option[FDeadline] to match the signature of `unmarshalDeadline` from `finagle-http`?